### PR TITLE
Docker local-testnet scripts

### PR DESCRIPTION
**Pull request by cristure**
## Reasoning behind the pull request
- Scipt to setup a local testnet with docker.
  
## Proposed changes
- Added scripts to setup local testnet.

## Testing procedure
- Run scripts/docker-testnet/start.sh
- Inspect the created containers.

## Pre-requisites
- Docker installed.

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: scripts/docker-testnet/helpers.sh [hunk 1]
```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env bash
```
Comment (iulianpascalau) [inline scripts/docker-testnet/helpers.sh:1]: why not rename this to "functions.sh" ?

File: scripts/docker-testnet/helpers.sh [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    # Truncate config.toml before the [[Observers]] list
+    sed -i -n '/\[\[Observers\]\]/q;p' config_edit.toml
+
+    if [ "$SHARD_OBSERVERCOUNT" -le 0 ]; then
+        generateProxyValidatorListDocker config_edit.toml
+    else
+        generateProxyObserverListDocker config_edit.toml
+    fi
+
+    cp config_edit.toml config.toml
```
Comment (iulianpascalau) [inline scripts/docker-testnet/helpers.sh:85]: `mv` directly and remove next line?

File: scripts/docker-testnet/start.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+prepareFolders
+
+buildConfigGenerator
+
+generateConfig
+
+copyConfig
+
+copySeednodeConfig
```
Comment (iulianpascalau) [inline scripts/docker-testnet/start.sh:22]: interesting, so you actually use a lot of the existing testnet scripts configuration functions :thinking:

Comment (cristure) [reply scripts/docker-testnet/start.sh:22]: yep, it's mostly the same configuration. The only difference in configuration are the ports and the ip addresses.

File: scripts/docker-testnet/helpers.sh (hunk 1 above)
Comment (cristure) [reply scripts/docker-testnet/helpers.sh:1]: done.

File: scripts/docker-testnet/start.sh (hunk 3 above)
Comment (iulianpascalau) [reply scripts/docker-testnet/start.sh:22]: right, I've noticed that: the first question was: how was possible to function if all nodes run on the same ports. But then I've saw the IPs.

**Review by iulianpascalau — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by AdoAdoAdo — DISMISSED** [review]

**Review by iulianpascalau — APPROVED** [review]
will test this ASAP

File: scripts/docker-testnet/build.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+pushd ../..
```
Comment (ssd04) [inline scripts/docker-testnet/build.sh:1]: add `#!/usr/bin/env bash` here also? i think pushd is bash specific

Comment (ssd04) [reply scripts/docker-testnet/build.sh:1]: is this script intended to be run separately?

File: scripts/docker-testnet/functions.sh [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  # this variable is used to keep track of the allocated IP addresses in the network, by removing the last byte
+  # of the DOCKER_NETWORK_SUBNET. One can consider this the host network address without the last byte at the end.
+  export NETWORK_ADDRESS=$(echo "$DOCKER_NETWORK_SUBNET" | rev | cut -d. -f2- | rev)
+}
+
+startSeedNode() {
+  docker run -d --name seednode -v ${TESTNETDIR}/seednode/config:/go/mx-chain-go/cmd/seednode/config \
+  --network ${DOCKER_NETWORK_NAME} \
+  seednode:dev \
+  --rest-api-interface=0.0.0.0:10000
```
Comment (ssd04) [inline scripts/docker-testnet/functions.sh:31]: indentation here?

File: scripts/docker-testnet/functions.sh [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+   done
+ done
+
+ for ((i = 0; i < META_OBSERVERCOUNT; i++)); do
+    KEY_INDEX=$((TOTAL_NODECOUNT - observerIdx - 1))
+
+    docker run -d --name "observer${observerIdx}-${NETWORK_ADDRESS}.${IP_HOST_BYTE}-10200-metachain" \
+        -v $TESTNETDIR/node/config:/go/mx-chain-go/cmd/node/config \
+        --network ${DOCKER_NETWORK_NAME} \
+        node:dev \
```
Comment (ssd04) [inline scripts/docker-testnet/functions.sh:64]: please use same indentation pattern for the entire script and remove unnecessary empty lines

File: scripts/docker-testnet/functions.sh [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+  else
+    cd $(dirname $MULTIVERSXDIR)
+
+    git clone git@github.com:multiversx/mx-chain-deploy-go.git || true
+    git clone git@github.com:multiversx/mx-chain-proxy-go.git || true
+  fi
+}
+
+createDockerNetwork() {
+  docker network create -d bridge --subnet=${DOCKER_NETWORK_SUBNET} ${DOCKER_NETWORK_NAME}
```
Comment (ssd04) [inline scripts/docker-testnet/functions.sh:20]: maybe check if `DOCKER_NETWOR_NAME` already exists

File: scripts/docker-testnet/clean.sh [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+export MULTIVERSXTESTNETSCRIPTSDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+source "$MULTIVERSXTESTNETSCRIPTSDIR/variables.sh"
+
+echo "Stopping all containers..."
+docker stop $(docker ps -a -q)
+
+echo "Removing all containers..."
+docker container prune -f
```
Comment (ssd04) [inline scripts/docker-testnet/clean.sh:13]: the scripts will most probably be used in development environment where there might be also other local containers running, i don't think it's fine to stop and prune all containers, we should stop only the intended containers

**Review by ssd04 — CHANGES_REQUESTED** [review]

File: scripts/docker-testnet/build.sh (hunk 4 above)
Comment (cristure) [reply scripts/docker-testnet/build.sh:1]: yes, it is. I could have added the build step inside the `./start.sh` script, but it was easier for me to debug, since it didn't build the images at every run. Don't have a strong opinion on this, I can add it to the wrapper script instead. I was planning on adding a `README.md` nevertheless.

File: scripts/docker-testnet/functions.sh (hunk 5 above)
Comment (cristure) [reply scripts/docker-testnet/functions.sh:31]: fixed.

File: scripts/docker-testnet/functions.sh (hunk 6 above)
Comment (cristure) [reply scripts/docker-testnet/functions.sh:64]: I hope it's fixed.

File: scripts/docker-testnet/functions.sh (hunk 7 above)
Comment (cristure) [reply scripts/docker-testnet/functions.sh:20]: added `set -u`. Unbound variables will throw an error now.

File: scripts/docker-testnet/build.sh (hunk 4 above)
Comment (iulianpascalau) [reply scripts/docker-testnet/build.sh]: I think it is better that we keep the build separate from the run script. Not a strong opinion, thus.

File: scripts/docker-testnet/variables.sh [hunk 9]
```diff
@@ -114,32 +108,28 @@ export PORT_ORIGIN_VALIDATOR_REST="9500"
...
-
-
-if [ "$TESTNETMODE" == "debug" ]; then
-  LOGLEVEL="*:DEBUG,api:INFO"
-fi
-
-if [ "$TESTNETMODE" == "trace" ]; then
-  LOGLEVEL="*:TRACE"
-fi
+## Use tmux or not. If set to 1, only 2 terminal windows will be opened, and
```
Comment (iulianpascalau) [inline scripts/docker-testnet/variables.sh:111]: commented code. Can use a local.sh file instead. We have exceptions for it in the .gitignore

File: scripts/docker-testnet/functions.sh [hunk 10]
```diff
@@ -6,14 +6,18 @@
...
 cloneRepositories() {
-  if [[ -n $CI_RUN ]]; then
-    echo "Repositories have been cloned in the CI"
-  else
-    cd $(dirname $MULTIVERSXDIR)
-
-    git clone git@github.com:multiversx/mx-chain-deploy-go.git || true
-    git clone git@github.com:multiversx/mx-chain-proxy-go.git || true
-  fi
+  cd $(dirname $MULTIVERSXDIR)
```
Comment (iulianpascalau) [inline scripts/docker-testnet/functions.sh:9]: Why this change? Does this function will **always** clone the proxy and deploy repos?

File: scripts/docker-testnet/functions.sh (hunk 7 above)
Comment (ssd04) [reply scripts/docker-testnet/functions.sh:24]: here I was referring to the fact that if the network is already created, it will fail because it is not able to create docker network again

File: scripts/docker-testnet/README.md [hunk 11]
```diff
@@ -0,0 +1,10 @@
+# Setting up a local-testnet with Docker
+
+First and foremost, one needs to build the **seednode** & **node** images. Hence, the **_build.sh_**
+script is provided. This can be done, by invoking the script or building the images manually.
+
+```
+./build.sh  # (Optional) Can be ignored if you already have the images stored in the local registry.
+./start.sh  # Will start the local-testnet.
+./clean.sh  # Will stop and remove the containers related to the local-testnet.
```
Comment (ssd04) [inline scripts/docker-testnet/README.md:9]: i think that a good flow here is to have also a `stop` step, in order to be able to have the following development flow:
- initially: `build`
- multiple times: `start` , `stop`
- and finally: `clean`

i think is not very good to have to clean everything on each run, something we just wan't to stop the squad and start it again, with the same/setup configuration, later on

Comment (ssd04) [reply scripts/docker-testnet/README.md:9]: this is also related to the docker create comment

File: scripts/docker-testnet/variables.sh (hunk 9 above)
Comment (cristure) [reply scripts/docker-testnet/variables.sh:111]: removed altogether. not needed in our context.

File: scripts/docker-testnet/functions.sh (hunk 10 above)
Comment (cristure) [reply scripts/docker-testnet/functions.sh:9]: yes, it will. But if the repository already exists, it won't clone them again.

File: scripts/docker-testnet/README.md (hunk 11 above)
Comment (cristure) [reply scripts/docker-testnet/README.md:9]: This intended workflow adds some complexity. The `start.sh` uses the `docker run` command. If you stop all containers, they will only start again with`docker start` instead, otherwise it will throw an error. I don't see much of a benefit to this behaviour, as I need to persists all the `containerIds/containerNames` between stopping and restarting. One can achieve this manually by leveraging already available commands in the `clean.sh` script.

File: scripts/docker-testnet/functions.sh (hunk 10 above)
Comment (iulianpascalau) [reply scripts/docker-testnet/functions.sh:9]: confirmed that is working properly :+1: 
```
git clone git@github.com:multiversx/mx-chain-deploy-go.git || true
fatal: destination path 'mx-chain-deploy-go' already exists and is not an empty directory.
```

**Review by iulianpascalau — DISMISSED** [review]
tested locally with system-tests-go and worked perfectly

File: scripts/docker-testnet/start.sh [hunk 12]
```diff
@@ -2,42 +2,17 @@
...
+    echo "File $file_path not found."
+    exit 1
 fi
 
+# Read the file line by line
+while IFS= read -r line; do
+    docker start $line
+done < "$file_path"
+
+rmdir ./tmp
```
Comment (iulianpascalau) [inline scripts/docker-testnet/start.sh:18]: add empty line?

File: scripts/docker-testnet/stop.sh [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+export CONTAINER_IDS=$(docker network inspect -f '{{range $k, $v := .Containers}}{{printf "%s\n" $k}}{{end}}' "$DOCKER_NETWORK_NAME")
+
+mkdir -p ./tmp
+
+# Stop each container
+echo "Stopping containers..."
+for CONTAINER_ID in $CONTAINER_IDS; do
+    docker stop "$CONTAINER_ID"
+    echo "$CONTAINER_ID" >> ./tmp/stopped_containers
+done
```
Comment (iulianpascalau) [inline scripts/docker-testnet/stop.sh:21]: add empty line?

**Review by iulianpascalau — DISMISSED** [review]

**Review by ssd04 — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]
